File: src/leetcode/decode-xored-array/decode-xored-array.js
/**
 * Leetcode #1720
 * Difficulty: Easy
 * https://leetcode.com/problems/decode-xored-array/submissions/
 */
/**
 * @param {number[]} encoded
 * @param {number} first
 * @return {number[]}
 */
export const decode = (encoded, first) => {
  const result = [first];

  for (let i = 0; i < encoded.length; i += 1) {
    result.push(encoded[i] ^ result[i]);
  }

  return result;
};

File: src/leetcode/decode-xored-array/decode-xored-array.rs
impl Solution {
    pub fn decode(encoded: Vec<i32>, first: i32) -> Vec<i32> {
        let mut ans = vec![first];

        for (i, &value) in encoded.iter().enumerate() {
            let idx: usize = i;
            ans.push(encoded[i] ^ ans[i]);
        }

        ans
    }
}

// impl Solution {
//     pub fn decode(encoded: Vec<i32>, first: i32) -> Vec<i32> {
//         std::iter::once(&first)
//             .chain(encoded.iter())
//             .scan(0, |xored, &x| {
//                 *xored ^= x;
//                 Some(*xored)
//             })
//             .collect::<Vec<i32>>()
//     }
// }